openshift/jenkins/Jenkinsfile

```diff
@@ -0,0 +1,323 @@
+#!groovy
+
+// ------------------
+// Pipeline Variables
+// ------------------
+def cli
+def notify
+
+// --------------------
+// Declarative Pipeline
+// --------------------
+pipeline {
+  agent none
+  options {
+    disableResume()
+    buildDiscarder(logRotator(numToKeepStr: "5")) // keep 5 builds only
+  }
+  environment {
+    CI = "true"
+    PROJECT_NAMESPACE = "jcxjin"
+    APP_NAME = "pims"
+    BUILD_ENV = "tools"
+    DEPLOY_TO = "dev"
+    DEPLOY_TO_UPPER = DEPLOY_TO.toUpperCase()
+    PATHFINDER_URL = "pathfinder.gov.bc.ca"
+    VANITY_URL="https://pims-dev.pathfinder.gov.bc.ca/"
+
+    WAIT_TIMEOUT = 10 // minutes
+
+    // enable pipeline verbose debug output if greater than 0
+    DEBUG_OUTPUT = "false"
+
+    // monorepo folder structure
+    API_CONTEXT_DIRECTORY = "backend"
+    FE_CONTEXT_DIRECTORY = "frontend"
+
+    TOOLS_PROJECT = "${PROJECT_NAMESPACE}-${BUILD_ENV}"
+    DEV_PROJECT = "${PROJECT_NAMESPACE}-${DEPLOY_TO}"
+  }
+  stages {
+    stage("Initialize") {
+      agent any
+      steps {
+        script {
+          // load supporting functions from external script files
+          cli = load "./openshift/jenkins/common.groovy"
+          notify = load "./openshift/jenkins/notify.groovy"
+
+          // ensure required secrets and credentials are available in the CI environment
+          // [add more secrets here if needed...]
+          sh "oc extract secret/rocket-chat-secrets --to=${env.WORKSPACE} --confirm"
+
+          // initialize CI environment values
+          env.ROCKET_DEPLOY_WEBHOOK = readFile("rocket-deploy-webhook").trim()
```

Comment:
I'm curious how this works?

Reply:
The secret management portion will need to be documented. I have followed best practices from other teams at the Exchange Lab. A subsequent PR could improve the documentation side of things

Reply:
But the TL;DR is:

* A secret is created with Rocket Chat's webhook URL; ie https://rc_server/hooks/<randomToken>
* I named the secret "rocket-chat-secrets" and it lives in the **tools** project
* Within the Jenkins pipeline, we have access to the Openshift CLI authenticated through a Jenkins service account (that's how the `openshift.withCluster()` etc works too)
* We can issue any `oc` command we want within the pipeline. In this case, we extract the secret as file under the current folder. This file will contain the secret as a string.
* We finally read the contents of the file into a pipeline variable for use further down below